code/xcvm/lib/core/src/asset.rs

```diff
@@ -402,3 +402,37 @@ impl<T> From<Funds<T>> for Vec<(u128, T)> {
 			.collect()
 	}
 }
+
+pub fn generate_asset_id(network_id: NetworkId, protocol_id: u32, nonce: u64) -> AssetId {
+	[
+		network_id.0.to_be_bytes().to_vec(),
+		protocol_id.to_be_bytes().to_vec(),
+		nonce.to_be_bytes().to_vec(),
+	]
+	.concat()
+	.try_into()
+	.map(u128::from_be_bytes)
+	.expect("[u8; 8] + bytes(u64) = [u8; 16]")
+	.into()
```

Comment:
There’s no need to allocate a Vector or for all the try_into and
except.  Just do simple arithmetic:
```suggestion
	AssetId::from((u128::from(network_id) << 96) |
		(u128::from(protocol_id) << 64) |
		(u128::from(nonce))
```